File: linked-list/js/Single-linked-list.js
class Node {
  constructor(value, next = null) {
    this.value = value;
    this.next = next;
  }
}

class LinkedList {
  constructor(value) {
    let node = new Node(value);
    this.head = this.tail = node;
    this.length = 1;
  }

  append(value) {
    let node = new Node(value);
    this.tail.next = node;
    this.tail = node;
    this.length++;
  }

  prepend(value) {
    let node = new Node(value);
    node.next = this.head;
    this.head = node;
    this.length++;
  }

  appendAtPosition(value, index) {
    if (index < 1 || index > this.length + 1) {
      console.log("Invalid index");
      return;
    }

    if (index === 1) {
      this.prepend(value);
      return;
    }

    if (index === this.length + 1) {
      this.append(value);
      return;
    }

    let node = new Node(value);
    let previousNode = this.findNode(index - 1);
    const temp = previousNode.next;
    previousNode.next = node;
    node.next = temp;
    this.length++;
  }

  findNode(index) {
    if (index < 1 || index > this.length) {
      console.log("Index out of bounds");
      return null;
    }

    let data = this.head;
    let count = 1;

    while (data) {
      if (count === index) {
        break;
      }
      data = data.next;
      count++;
    }
    return data;
  }

  updateNode(value, index) {
    let previousNode = this.findNode(index);
    previousNode.value = value;
  }

  deleteNode(index) {
    if (index < 1 || index > this.length) {
      console.log("Invalid index");
      return;
    }

    if (index === 1) {
      this.head = this.head.next;
      this.length--;
      if (this.length === 0) this.tail = null;
      return;
    }

    let previousNode = this.findNode(index - 1);

    if (index === this.length) {
      this.tail = previousNode;
      this.tail.next = null;
    } else {
      previousNode.next = previousNode.next.next;
    }

    this.length--;
  }

  print() {
    let data = this.head;

    while (data) {
      console.log(data.value);
      data = data.next;
    }
  }
}

const list = new LinkedList(10);
list.append(20);
list.prepend(5);
list.updateNode(15, 2);
list.appendAtPosition(25, 4);
list.deleteNode(1);
list.deleteNode(3);
list.append(50);
list.deleteNode(2)
list.print();
console.log(list);

File: linked-list/py/Single-linked-list.py

class Node:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next
 



class Single_linked_list:
    def __init__(self, val=None):
        self.head = self.tail = Node(val) if val is not None else None
        self.len = 1 if val is not None else 0


    def append_value(self, val):
        node = Node(val)
        if self.head is None:  
            self.head = self.tail = node
        else:
            self.tail.next = node
            self.tail = node
        self.len += 1 


    def prepend_value(self,val):
        node = Node(val)
        if self.head is None:
            self.head = self.tail = node
        else:
            node.next = self.head
            self.head = node

        self.len += 1


    def append_at_position(self,val,index):
        if index < 1 or index > self.len + 1:
          print("invalid index")
          return
        
        if index == 1:
          self.prepend_value(val)
          return 
        
        if index == self.len + 1:
          self.append_value(val)
          return
        
        node = Node(val)
        previous_node = self.find_node(index-1)
        temp = previous_node.next
        previous_node.next = node
        node.next = temp
        self.len += 1        


    def find_node(self,index):

        if index < 1 or index > self.len:
          print("index is out of range")
          return
        
        data = self.head
        num = 1

        while data:
          if num == index:
            break
          
          num +=1
          data = data.next
        
        return data
    

    def update_value(self,val,index):
     
      if index < 1 or index > self.len:
        print("index is out of range")
        return 
     
      node = self.find_node(index)
      node.val = val
    

    def print_list(self):
       node = self.head
       while node:
        print(node.val, end=" -> ")
        node = node.next
       print("None")






my_list = Single_linked_list()
my_list.append_value(10)
my_list.append_value(20)
my_list.prepend_value(5)
# print("find node",my_list.find_node(4))
my_list.append_at_position(1,1)
my_list.append_at_position(25,5)
my_list.append_at_position(15,4)
my_list.update_value(1555,4)
my_list.print_list()
print("head",my_list.head.val)
print("tail",my_list.tail.val)
print("len",my_list.len)

